src/main/java/uk/gov/pay/directdebit/webhook/gocardless/services/handlers/SendEmailsForGoCardlessEventsHandler.java

```diff
@@ -0,0 +1,91 @@
+    
+    private static Predicate<GoCardlessEvent> byValidEventsForMandateResourceTypes =
+            event -> MANDATES.equals(event.getResourceType());
+    
+    private static Predicate<GoCardlessEvent> byValidEventsForPaymentsAndPayoutsResourceTypes =
+            event -> Set.of(PAYMENTS, PAYOUTS).contains(event.getResourceType()) &&
+                    Set.of("created", "submitted", "confirmed", "failed", "paid_out", "paid")
+                            .contains(event.getAction().toLowerCase());
+
+    private final MandateQueryService mandateQueryService;
+    private final UserNotificationService userNotificationService;
+    private final PaymentQueryService paymentQueryService;
+
+    @Inject
+    public SendEmailsForGoCardlessEventsHandler(MandateQueryService mandateQueryService,
+                                                UserNotificationService userNotificationService,
+                                                PaymentQueryService paymentQueryService) {
+        this.mandateQueryService = mandateQueryService;
+        this.userNotificationService = userNotificationService;
+        this.paymentQueryService = paymentQueryService;
+    }
+
+    public void handle(List<GoCardlessEvent> events) {
+        events.stream().filter(byValidEventsForMandateResourceTypes).forEach(e -> handleMandate(e));
+        events.stream().filter(byValidEventsForPaymentsAndPayoutsResourceTypes).forEach(e -> handlePayment(e));
+    }
+
+    private void handlePayment(GoCardlessEvent event) {
+        logEvent(event);
+        GoCardlessPaymentId goCardlessPaymentId = event.getLinksPayment()
+                .orElseThrow(() -> new GoCardlessEventHasNoPaymentIdException(event.getGoCardlessEventId()));
+
+        Payment payment = paymentQueryService.findByGoCardlessPaymentIdAndOrganisationId(goCardlessPaymentId, event.getLinksOrganisation())
+                .orElseThrow(() -> new PaymentNotFoundException(goCardlessPaymentId, event.getLinksOrganisation()));
+
+        if (event.getAction().equalsIgnoreCase("failed")) {
+            userNotificationService.sendPaymentFailedEmailFor(payment);
+        }
+    }
+
+    private void handleMandate(GoCardlessEvent event) {
+        logEvent(event);
+        switch (event.getAction().toLowerCase()) {
+            case "cancelled": 
+                userNotificationService.sendMandateCancelledEmailFor(getMandate(event)); 
+                break;
+            case "failed": 
+                userNotificationService.sendMandateFailedEmailFor(getMandate(event));
+        }
+    }
+
+    private Mandate getMandate(GoCardlessEvent event) {
+        return mandateQueryService.findByGoCardlessMandateIdAndOrganisationId(
+                    event.getLinksMandate().orElseThrow(() -> new GoCardlessEventHasNoMandateIdException(event.getGoCardlessEventId())),
+                    event.getLinksOrganisation());
+    }
+
+    private static void logEvent(GoCardlessEvent event) {
```

Comment:
I think this logging has outlived its usefulness now we don’t handle an event in one go. We should already have logging for persisting events, changing statuses due to events and sending emails because of events.

Reply:
ok. deleted.